README.md
ATTENTION: Development of this project has moved to: https://github.com/bnotions/twoscoops-chef



twoscoops Cookbook
==================
TODO: Enter the cookbook description here.

e.g.
This cookbook makes your favorite breakfast sandwhich.

Requirements
------------
TODO: List your cookbook requirements. Be sure to include any requirements this cookbook has on platforms, libraries, other cookbooks, packages, operating systems, etc.

e.g.
#### packages
- `toaster` - twoscoops needs toaster to brown your bagel.

Attributes
----------
TODO: List you cookbook attributes here.

e.g.
#### twoscoops::default
<table>
  <tr>
    <th>Key</th>
    <th>Type</th>
    <th>Description</th>
    <th>Default</th>
  </tr>
  <tr>
    <td><tt>['twoscoops']['bacon']</tt></td>
    <td>Boolean</td>
    <td>whether to include bacon</td>
    <td><tt>true</tt></td>
  </tr>
</table>

Usage
-----
#### twoscoops::default
TODO: Write usage instructions for each cookbook.

e.g.
Just include `twoscoops` in your node's `run_list`:

```json
{
  "name":"my_node",
  "run_list": [
    "recipe[twoscoops]"
  ]
}
```

Contributing
------------
TODO: (optional) If this is a public cookbook, detail the process for contributing. If this is a private cookbook, remove this section.

e.g.
1. Fork the repository on Github
2. Create a named feature branch (like `add_component_x`)
3. Write you change
4. Write tests for your change (if applicable)
5. Run the tests, ensuring they all pass
6. Submit a Pull Request using Github

License and Authors
-------------------
Authors: TODO: List authors


CHANGELOG.md
twoscoops CHANGELOG
===================

This file is used to list changes made in each version of the twoscoops cookbook.

0.1.0
-----
- [your_name] - Initial release of twoscoops

- - -
Check the [Markdown Syntax Guide](http://daringfireball.net/projects/markdown/syntax) for help with Markdown.

The [Github Flavored Markdown page](http://github.github.com/github-flavored-markdown/) describes the differences between markdown on github and standard markdown.


recipes/test.rb
#
# Cookbook Name:: twoscoops
# Recipe:: test
#
# Copyright 2013, YOUR_COMPANY_NAME
#
# All rights reserved - Do Not Redistribute
#

include_recipe "twoscoops::local"

execute "pip-install-test-requirements" do
  cwd "/vagrant"
  command "pip install -r requirements/test.txt"
end

execute "django-jenkins" do
  cwd "/vagrant/#{node['twoscoops']['project_name']}"
  command "./manage.py jenkins --settings=#{node['twoscoops']['project_name']}.settings.test"
end


recipes/deploy.rb
#
# Cookbook Name:: twoscoops
# Recipe:: deploy
#
# Copyright 2013, YOUR_COMPANY_NAME
#
# All rights reserved - Do Not Redistribute
#

secret = Chef::EncryptedDataBagItem.load_secret("/tmp/encrypted_data_bag_secret")
github_keys = Chef::EncryptedDataBagItem.load("github-deploy", node['twoscoops']['application_name'], secret)

include_recipe "twoscoops::database"
include_recipe "twoscoops::webserver"

user "deploy" do
  home node['twoscoops']['application_deploy_path']
  shell "/bin/bash"
  supports :manage_home => true
end

application node['twoscoops']['application_name'] do
  packages ["git-core"]
  path "#{node['twoscoops']['application_deploy_path']}/#{node['twoscoops']['application_name']}"
  repository node['twoscoops']['application_repository']
  revision node['twoscoops']['application_revision']
  deploy_key github_keys['private_key']
  owner "deploy"
  group "deploy"

  migrate true
  migration_command "echo 'migrate!'"
  symlink_before_migrate ({
    "settings/database.py" => "#{node['twoscoops']['project_name']}/#{node['twoscoops']['project_name']}/settings/database.py",
    "app_environment.sh" => "#{node['twoscoops']['project_name']}/app_environment.sh" 
  })
  before_migrate do
    directory "#{shared_path}/settings" do
      recursive true
    end

    template "#{shared_path}/settings/database.py" do
      source "database.py.erb"
      mode 00644
    end

    template "#{shared_path}/app_environment.sh" do
      source "environment.sh.erb"
      mode 00755
    end
  end

  restart_command do
    supervisor_service "uwsgi" do
      action :restart
    end
  end
  before_restart do
    execute "pip-install-requirements" do
      cwd "#{shared_path}/../current"
      command "pip install -r requirements.txt"
    end

    bash "django-syncdb" do
      cwd "#{shared_path}/../current/#{node['twoscoops']['project_name']}"
      code <<-EOF
        source app_environment.sh
        ./manage.py syncdb --noinput
      EOF
    end

    bash "django-createcachetable" do
      cwd "#{shared_path}/../current/#{node['twoscoops']['project_name']}"
      code <<-EOF
        source app_environment.sh
        ./manage.py createcachetable application_cache
      EOF
      only_if do
        con = PGconn.connect("host=localhost user=postgres password=#{node['postgresql']['password']['postgres']} dbname=#{node['twoscoops']['application_name']}")
        res = con.exec("SELECT count(*) FROM information_schema.tables WHERE table_name = 'application_cache'")
        res.entries[0]['count'] == 0
      end
    end

    directory "/tmp/twoscoops/fixtures" do
      recursive true
      action :create
      mode 00755
    end

    template "/tmp/twoscoops/fixtures/createsuperuser.json" do
      source "createsuperuser.json.erb"
    end

    bash "django-createsuperuser" do
      cwd "#{shared_path}/../current/#{node['twoscoops']['project_name']}"
      code <<-EOF
        source app_environment.sh
        ./manage.py loaddata /tmp/twoscoops/fixtures/createsuperuser.json
      EOF
    end

    bash "django-migrate" do
      cwd "#{shared_path}/../current/#{node['twoscoops']['project_name']}"
      code <<-EOF
        source app_environment.sh
        ./manage.py migrate
      EOF
    end

    bash "django-collectstatic" do
      cwd "#{shared_path}/../current/#{node['twoscoops']['project_name']}"
      code <<-EOF
        source app_environment.sh
        ./manage.py collectstatic --noinput
      EOF
    end
  end
end



recipes/database.rb
#
# Cookbook Name:: twoscoops
# Recipe:: database
#
# Copyright 2013, YOUR_COMPANY_NAME
#
# All rights reserved - Do Not Redistribute
#

if node['twoscoops']['database']['engine'] == 'django.db.backends.postgresql_psycopg2'
  include_recipe "database::postgresql"

  database_connection_info = {
    :host => node['twoscoops']['database']['host'],
    :username => 'postgres',
    :password => node['postgresql']['password']['postgres']
  }

  postgresql_database node['twoscoops']['application_name'] do
    connection database_connection_info
    provider Chef::Provider::Database::Postgresql
    action :create
  end

  postgresql_database_user node['twoscoops']['database']['username'] do
    connection database_connection_info
    password node['twoscoops']['database']['password']
    action :create
  end

  postgresql_database_user node['twoscoops']['database']['username'] do
    connection database_connection_info
    database_name node['twoscoops']['application_name']
    privileges [:all]
    action :grant
  end
end


recipes/webserver.rb
#
# Cookbook Name:: twoscoops
# Recipe:: production
#
# Copyright 2013, YOUR_COMPANY_NAME
#
# All rights reserved - Do Not Redistribute
#

include_recipe "nginx"

template "#{node['nginx']['dir']}/sites-enabled/#{node['twoscoops']['application_name']}" do
  source "nginx.conf.erb"
  mode 0440
  owner "root"
  group "root"  
end

nginx_site node['twoscoops']['application_name'] do
  action :enable
end

include_recipe "supervisor"
include_recipe "uwsgi"

supervisor_service "uwsgi" do
  user "www-data"
  command "/usr/local/bin/uwsgi --uwsgi-socket :8080 --wsgi-file #{node['twoscoops']['project_name']}/wsgi.py --touch-reload=/tmp/uwsgi_restart.txt"
  autostart true
  directory "#{node['twoscoops']['application_deploy_path']}/#{node['twoscoops']['application_name']}/current/#{node['twoscoops']['project_name']}"
  stdout_logfile "/var/log/uwsgi.log"
  stderr_logfile "/var/log/uwsgi_error.log"
  environment "SECRET_KEY" => node['twoscoops']['secret_key'],
              "DJANGO_SETTINGS_MODULE" => "#{node['twoscoops']['project_name']}.settings.#{node['twoscoops']['application_environment']}"
  action :enable
end


recipes/local.rb
#
# Cookbook Name:: twoscoops
# Recipe:: development
#
# Copyright 2013, YOUR_COMPANY_NAME
#
# All rights reserved - Do Not Redistribute
#

include_recipe "twoscoops::database"

directory "/vagrant/logs" do
  action :create
  mode 00755
end

template "/vagrant/#{node['twoscoops']['project_name']}/#{node['twoscoops']['project_name']}/settings/database.py" do
  source "database.py.erb"
  mode 00644
end

execute "pip-install-requirements" do
  cwd "/vagrant"
  command "pip install -r requirements/local.txt"
end

execute "django-syncdb" do
  cwd "/vagrant/#{node['twoscoops']['project_name']}"
  command "python manage.py syncdb --noinput"
end

directory "/tmp/twoscoops/fixtures" do
  recursive true
  action :create
  mode 00755
end

template "/tmp/twoscoops/fixtures/createsuperuser.json" do
  source "createsuperuser.json.erb"
end

execute "django-createsuperuser" do
  cwd "/vagrant/#{node['twoscoops']['project_name']}"
  command "python manage.py loaddata /tmp/twoscoops/fixtures/createsuperuser.json"
end

execute "django-migrate" do
  cwd "/vagrant/#{node['twoscoops']['project_name']}"
  command "python manage.py migrate"
end

include_recipe "supervisor"

supervisor_service "django" do
  command "python manage.py runserver 0.0.0.0:8080"
  autostart true
  directory "/vagrant/#{node['twoscoops']['project_name']}"
  stdout_logfile "/vagrant/logs/django.log"
  stderr_logfile "/vagrant/logs/django_error.log"
  action :enable
end

